pkg/gameservers/portallocator.go

```diff
@@ -156,33 +156,57 @@ func (pa *PortAllocator) syncPorts(key string) error {
 // Allocate assigns a port to the GameServer and returns it.
 // Return ErrPortNotFound if no port is allocatable
 func (pa *PortAllocator) Allocate(gs *v1alpha1.GameServer) (*v1alpha1.GameServer, error) {
-	if gs.Spec.PortPolicy != v1alpha1.Dynamic {
-		return gs, errors.Errorf("Port policy of %s is not supported for port allocation", gs.Spec.PortPolicy)
-	}
-
 	pa.mutex.Lock()
 	defer pa.mutex.Unlock()
-	for _, n := range pa.portAllocations {
-		for p, taken := range n {
-			if !taken {
-				n[p] = true
-				pa.gameServerRegistry[gs.ObjectMeta.UID] = true
-				gs.Spec.HostPort = p
-				return gs, nil
+
+	type pn struct {
+		pa   portAllocation
+		port int32
+	}
+
+	// we only want this to be called inside the mutex lock
+	// so let's define the function here so it can never be called elsewhere.
+	// Also the return gives an escape from the double loop
+	findOpenPorts := func(amount int) []pn {
```

Comment:
@enocom would love to get your take on what I'm doing here.

Reply:
I've seen this pattern of declaring a function inside a method and then used in the stdlib. It's something I've been meaning to do more often myself. Very nice.